Feature map 63 location list endpoint

6mahdapihka9 opened the pull request:
created endpoint for location list with map functionality and popups on frontend
-_- tested - seems good to me

for Oleksandr Poberezhnyi: data for your popups stored in variable location (row 43 in map component), it's only ids and coords for further request

6mahdapihka9 committed:
location controller added

6mahdapihka9 committed:
Merge branch 'develop' of https://github.com/IT-Academy-Social-Projects-KRV/MapUA into feature_location_list_endpoint

 Conflicts:
	frontend/src/components/Map/index.tsx

6mahdapihka9 committed:
consts for testing

6mahdapihka9 committed:
Created state for zoom and locations that are showed on map.
Used map hook on zoom to fetch needed amount of locations on each zoom.
Todo: use map hook on drag map and fetch location depending on center position

6mahdapihka9 committed:
Fix for popup icon

6mahdapihka9 committed:
uuid package for key generation

6mahdapihka9 committed:
comments cleared, place by which data will be passed to backend changed to params (instead of req body)

6mahdapihka9 committed:
Merge branch 'develop' of https://github.com/IT-Academy-Social-Projects-KRV/MapUA into feature_MAP-63_location_list_endpoint

 Conflicts:
	backend/controllers/LocationsController.ts
	frontend/yarn.lock

6mahdapihka9 committed:
fixed endpoint bug with port and request param

6mahdapihka9 committed:
fetch location by bounds and center, added methods to handle zoom, drag and click, rewrote data in db

6mahdapihka9 committed:
Merge branch 'develop' of https://github.com/IT-Academy-Social-Projects-KRV/MapUA into feature_MAP-63_location_list_endpoint

 Conflicts:
	backend/.envExample
	backend/yarn.lock

6mahdapihka9 committed:
package bug fixed, added exception catcher in fetcher, code cleared a bit

6mahdapihka9 committed:
top 50 of requested data

romawht22 commented on frontend/src/components/Map/index.tsx:
```diff
@@ -13,12 +104,18 @@ function Map() {
       <TileLayer url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png" />
+      {locations.map((loc: lightLocationType) => (
+        <Marker position={loc.coordinates} key={v4()}>
+          <Popup>
+            A pretty CSS3 popup.
+            <br />
+            Easily customizable.
+          </Popup>
```
we have our own custom popup,delete this one

romawht22 commented on frontend/src/index.tsx:
```diff
@@ -9,10 +9,21 @@ import React from 'react';
 import reportWebVitals from './reportWebVitals';
 
+// fix for popup icon
+const DefaultIcon = L.icon({
+  iconUrl: icon,
+  shadowUrl: iconShadow
+});
+L.Marker.prototype.options.icon = DefaultIcon;
```
add this image as prop to a <Marker/> in Map component

steadyGuy commented on backend/controllers/LocationsController.ts:
```diff
@@ -1,45 +1,81 @@
+import { Response, Request } from "express";
+import Location from "../models/Locations";
+// import { locations as locConst, locationType } from '../../frontend/src/constants/develop';
+// import {equal} from "assert";
```
remove

steadyGuy commented on frontend/src/components/Map/index.tsx:
```diff
@@ -1,9 +1,100 @@
+import { lightLocationType } from '../../constants/develop';
+import { fetchData } from '../../fetch/requests';
+
+type latlngType = {
+  lat: number;
+  lng: number;
+};
+
+type boundsType = {
+  _northEast: {
+    lat: number;
+    lng: number;
+  };
+  _southWest: {
+    lat: number;
+    lng: number;
+  };
+};
```
move to _types.ts_ file and export types from that file.
Also, I don't understand why u use underscore with northEast and _southWest variables

steadyGuy commented on frontend/src/fetch/requests.ts:
```diff
@@ -0,0 +1,18 @@
+export async function fetchData(url: string, option?: any) {
+  let status = 0;
+  let data = null;
+  try {
+    data = await fetch(url, option)
+      .then((res: any) => {
+        status = res.status;
+        return res;
+      })
+      .then((res: any) => res.json())
+      .catch((e: any) => {
+        console.error('Error has occurred: ', e);
+      });
+  } catch (e) {
+    console.log(e);
+  }
+  return { status, data };
+}
```
Move this file to utils, and notify a team about this global function for fetching data.

steadyGuy commented on backend/controllers/LocationsController.ts:
```diff
@@ -1,45 +1,81 @@
+        coordinates: l.coordinates,
+      }));
+
+      // if (!locations || locations.length) {
+      //   return res.status(400).json({ message: "Location doesn't exist" });
+      // }
```
why commented?

steadyGuy commented on backend/controllers/LocationsController.ts:
```diff
@@ -1,45 +1,81 @@
+          l.coordinates[0] < center[0] + height &&
+          l.coordinates[1] > center[1] - width &&
+          l.coordinates[1] < center[1] + width)
+      locations = locations.slice(0, (locations.length < 50)? locations.length : 50);
```
You shoud filter in request to mongoDb, not after we get all coordinates.

steadyGuy commented on backend/controllers/LocationsController.ts:
```diff
@@ -1,45 +1,81 @@
+      let locations = (await Location.find()).map(l => ({
+        _id: l._id,
+        coordinates: l.coordinates,
+      }));
```
apply filters here, in mongoose method find

steadyGuy commented on backend/controllers/LocationsController.ts:
```diff
@@ -1,45 +1,81 @@
+        const result = await newLocation.save(newLocation as any);
+        res.status(200).json(result);
+      } else {
+        res.status(500).json({ message: "Data is present" });
```
It's 400 code, not 500

6mahdapihka9 replied:
it's Andrii's part

steadyGuy commented on backend/routes/index.ts:
```diff
@@ -7,9 +7,11 @@ const router = express.Router();
 
 router.get("/users", UserController.getUsers);
 router.get("/locations/:id", LocationsController.getLocationById);
+router.get("/locations/search/:zoom/:center/:bounds", LocationsController.getLocationsByZoom);
```
Maybe, it's better to use req.query instead of req.params, as it's much easier on frontend to change query, rather than url (but it's just a remark, so it's up to you, change it or not).

6mahdapihka9 replied:
changed method to post, so it could be scaled in the future

steadyGuy commented on frontend/src/constants/develop.ts:
```diff
@@ -0,0 +1,379 @@
+  coordinates: [number, number];
+  photoSrc: string;
+};
+
```
inside types.ts

steadyGuy commented on frontend/src/components/Map/index.tsx:
```diff
@@ -1,9 +1,100 @@
+
+  useEffect(() => {
+    async function onZoom() {
+      const url = `http://localhost:3001/api/locations/search/${zoom}/${JSON.stringify(
```
put inside .env file (http://localhost:3001/api/) in frontend (variable name REACT_APP_API_URI)

steadyGuy commented on frontend/src/components/Map/index.tsx:
```diff
@@ -1,9 +1,100 @@
+    const map = useMapEvents({
+      zoom: e => {
+        setZoom(map.getZoom());
+        // eslint-disable-next-line no-underscore-dangle
```
=)

steadyGuy commented on frontend/src/components/Map/index.tsx:
```diff
@@ -1,9 +1,100 @@
+      },
+      click: () => null,
+      dragend: e => {
+        // eslint-disable-next-line no-underscore-dangle
```
=)

Yojuks replied:
what is it?

steadyGuy reviewed (commented):
Remove package-lock.json in frontend and backend folders.
Remove package.json from the root folder. This file exists only in backend/frontend folder.

6mahdapihka9 committed:
removed logs and comments, moved location filter into mongodb method

6mahdapihka9 committed:
Type objects moved to types.ts file.
Removed default popups.
Changed endpoint method, so it could be scaled later.
Removed unused comments in endpoint controller.
Modified fetch method moved to utils folder.

6mahdapihka9 committed:
got rid of useless useState hooks

Yojuks commented on backend/.envExample:
```diff
@@ -3,4 +3,4 @@ MONGO_PASSWORD=
 MONGO_DB=
 MONGO_HOSTNAME=
 PORT=3001
-HOST = http://localhost:3001
```
You want to add new line?)

6mahdapihka9 replied:
it's not mine row, it was merged into my branch after pulling from develop branch

6mahdapihka9 committed:
changed source of url default path to process env variable

6mahdapihka9 committed:
variable declaration fix

6mahdapihka9 committed:
Merge branch 'develop' of https://github.com/IT-Academy-Social-Projects-KRV/MapUA into feature_MAP-63_location_list_endpoint

 Conflicts:
	frontend/src/components/Map/index.tsx
	frontend/yarn.lock

6mahdapihka9 committed:
Merge branch 'develop' into feature_MAP-63_location_list_endpoint

steadyGuy reviewed (changes requested):
Come on) You did'nt remove package-lock.json and package.json on the root folder
![image](https://user-images.githubusercontent.com/41287687/164054910-3a72d6a1-0e2b-4060-9694-59045bbbfd51.png)

6mahdapihka9 committed:
Merge branch 'develop' of https://github.com/IT-Academy-Social-Projects-KRV/MapUA into feature_MAP-63_location_list_endpoint

 Conflicts:
	backend/yarn.lock

6mahdapihka9 committed:
removed packages from root folder

6mahdapihka9 committed:
Merge remote-tracking branch 'origin/feature_MAP-63_location_list_endpoint' into feature_MAP-63_location_list_endpoint
